### ISSUE TITLE
Clarify third-party licenses and license Windows builds under GPL due to `mutagen`

### ISSUE THREAD

**Issue by louie-github** [yt-dlp/yt-dlp#348]
## Checklist

- [ ] I'm reporting a broken site support issue
- [x] I've verified that I'm running yt-dlp version **2021.05.20**
- [ ] I've checked that all provided URLs are alive and playable in a browser
- [ ] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the bugtracker for similar bug reports including closed ones
- [x] I've read bugs section in FAQ

## Verbose log

N/A

## Description

As I was reading the README, I noticed this line:
> Although there are no other required dependencies, `ffmpeg` and `ffprobe` are highly recommended. Other optional dependencies are `sponskrub`, `AtomicParsley`, `mutagen`, `pycryptodome`, `phantomjs` and any of the supported external downloaders. Note that the windows releases are already built with the python interpreter, mutagen and pycryptodome included. 

I decided to check the licenses for each project, and got:
 - [ffmpeg (and ffprobe)](https://www.ffmpeg.org/legal.html): LGPLv2.1 or later / GPLv2 or later / non-free, depending on the build
 - [sponskrub](https://github.com/faissaloo/SponSkrub/blob/master/LICENCE.md): GPLv3 or later
 - [AtomicParsley](https://github.com/wez/atomicparsley/blob/master/COPYING): GPLv2 or later
 - [mutagen](https://github.com/quodlibet/mutagen/blob/master/COPYING): GPLv2 or later
 - [pycryptodome](https://github.com/Legrandin/pycryptodome/blob/master/LICENSE.rst): BSD 2-Clause (with some code in the public domain via Unlicense)
 - [phantomjs](https://github.com/ariya/phantomjs/blob/master/LICENSE.BSD): BSD 3-Clause

First of all, I recommend that we include a line stating that these optional dependencies may be distributed under different license terms than `yt-dlp`, and so, using them and/or redistributing them will require you to follow their specific licenses:

An example:
> Each optional dependency may be licensed under different terms than `yt-dlp`. If you choose to use these optional dependencies, you must agree to their license terms. Please refer to each dependency's repository or README file for more information.
>
> Additionally, if you wish to redistribute `yt-dlp` with any of these optional dependencies included, you must ensure that you follow all of your dependencies' license terms, such as distributing the whole software under the terms of the GPL if you link to a GPL library.

Additionally, because of this line pertaining to the Windows builds:

> Note that the windows releases are already built with the python interpreter, mutagen and pycryptodome included. 

I believe that the Windows builds now fall under the terms of the GPL since `mutagen` is licensed under the GPLv2 and we *include* it in the final binary, as well as using it in the code.

Additionally, we should include both a LICENSE file containing the text of the GPLv2 license (because of `mutagen`) and a
THIRD-PARTY-NOTICES file or similar that includes the text of the license for `pycryptodome` and for `Python` itself since we embed the interpreter into the executable.

PSF License and license stack: https://docs.python.org/3/license.html
FAQ on bundling Python: https://wiki.python.org/moin/PythonSoftwareFoundationLicenseFaq

> *If I bundle Python with my application, what do I need to include in my
> software and/or printed documentation?*
> 
> You must retain all copyright notices found in the code you are redistributing and include a copy of the PSF License and all of the other licenses in the Python "license stack" with the software distribution. The "license stack" is a result of the history of Python's development as described here (this page also lists all licenses in the stack).
> 
> Separate from the issue of copyright, the name “Python” is also a trademark. You should include the notice “"Python" is a registered trademark of the Python Software Foundation” in the appropriate part of your documentation or About box and place the “®” symbol after the first mention of “Python” in your documentation. (For example: “Python® is used as the scripting language for...”)

I'm not sure whether `yt-dlp` as a whole needs to be licensed under the GPL since it's not exactly clear what constitutes as "linking" and therefore making a "combined work", but I'd (personally) lean on the side of "no" since we disable the relevant features if the dependency is not present. The way we use some of these libraries isn't exactly deeply linking into them, but that's up for debate and out of the scope of this issue.

Though, I do finally suggest that we remove `mutagen` and `pycryptodome` from [requirements.txt](requirements.txt) or at least make them an optional dependency under another package like `yt-dlp[full]` or something which is listed as GPL-licensed rather than Public Domain just to be safe.

In summary, I have 4 suggestions:
1. Add text in the README saying that optional dependencies have different license terms
2. Add text in the README saying that redistributions which include optional dependencies may be licensed under different terms than the Unlicense
3. License the Windows builds under GPL and package LICENSE + THIRD-PARTY-NOTICES files along with the `.exe`
4. Move `mutagen` and `pycryptodome` requirements to another package like `yt-dlp[full]` with GPL license if possible

I apologize for the rather long write-up, and I am willing to help with this if necessary.

**Comment by pukkandan** [comment]
> 1. Add text in the README saying that optional dependencies have different license terms
> 2. Add text in the README saying that redistributions which include optional dependencies may be licensed under different terms than the Unlicense

> Separate from the issue of copyright, the name “Python” is also a trademark. You should include the notice “"Python" is a registered trademark of the Python Software Foundation” in the appropriate part of your documentation or About box and place the “®” symbol after the first mention of “Python” in your documentation. (For example: “Python® is used as the scripting language for...”)

These can be easily done

---

> and a THIRD-PARTY-NOTICES file or similar ... and for Python itself since we embed the interpreter into the executable.

The python interpretter is  not "bundled" into the release package in the conventional sense. As I understand, the "bootloader exception" in pyinstaller's licence already covers for this
```
Bootloader Exception
--------------------

In addition to the permissions in the GNU General Public License, the
authors give you unlimited permission to link or embed compiled bootloader
and related files into combinations with other programs, and to distribute
those combinations without any restriction coming from the use of those
files. (The General Public License restrictions do apply in other respects;
for example, they cover modification of the files, and distribution when
not linked into a combined executable.)
```

If it does not, then `youtube-dl` is also in violation of the license since they use `py2exe` which similarly embeds required portions of the python interpretter

---

> and a THIRD-PARTY-NOTICES file or similar that includes the text of the license for pycryptodome

The BSD-2 licence doesn't impose any restrictions on linking (similar to AGPL), so this is not necessary

---

> I believe that the Windows builds now fall under the terms of the GPL since mutagen is licensed under the GPLv2 and we include it in the final binary

This is in-fact an issue

> , as well as using it in the code.

We are not distributing it with our code, so it does not matter

> License the Windows builds under GPL and package LICENSE + THIRD-PARTY-NOTICES files along with the .exe

As I understand, adding a COPYING.txt to the releases is sufficient. Could you help me write one?

> Move mutagen and pycryptodome requirements to another package like yt-dlp[full] with GPL license if possible

No. The requirements.txt may contain either necessary or optional dependancies according to PyPi documents. Also, adding the libraries to requirements.txt does not violate any licenses

---

I am just a guy trying to improve on the youtube-dl project and make my modifications available to the public. I am honestly not interested in all this copyright bullshit. But I suppose it is a part of releasing a project 😩

**Comment by pukkandan** [comment]
Also, if there are any legal professionals here, I would like to hear your opinions on the matter

**Comment by louie-github** [comment]
> The python interpretter is not "bundled" into the release package in the conventional sense. As I understand, the "bootloader exception" in pyinstaller's licence already covers for this

Hmm, I can't really comment any further on this since I don't claim to understand things fully.  But, for now, I keep to my opinion that embedding Python would require us to place the text of the PSF License, until I research things further and see for myself. This is your call, though, of course.

> The BSD-2 licence doesn't impose any restrictions on linking (similar to AGPL), so this is not necessary

That is true, it does not impose any restrictions on linking. Though, I was referring to adding the license notice to the Windows binary package since we also include `pycryptodome` there, and the license states that "Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution."

> As I understand, adding a COPYING.txt to the releases is sufficient. Could you help me write one?

Sure, I'll try my best, but I should probably read the GPL and try to research some more to see what exactly we need to put in that file.

***Anyone who's worked with GPL code before, it would be great if you could provide some information or assistance.***

> We are not distributing it with our code, so it does not matter

I think that if your code absolutely cannot function without a GPL library (like, for example, calling the private interfaces of the code rather than a public API which theoretically could be implemented by a different program) it probably falls under the GPL, but, again, I am very much not sure on that.

To be clear, I agree that the way we include `mutagen` (where it is an optional dependency) should make it not matter unless you *include* it.

> No. The requirements.txt may contain either necessary or optional dependancies according to PyPi documents. Also, adding the libraries to requirements.txt does not violate any licenses

Oh, okay. That clears things up a bit.

> I am just a guy trying to improve on the youtube-dl project and make my modifications available to the public. I am honestly not interested in all this copyright bullshit. But I suppose it is a part of releasing a project 😩

I definitely understand. It's really annoying to have to deal with all the licenses and legal stuff that comes with using open-source software. A lot of the projects I look at seem to not really bother with it since it probably won't affect them if they're not a huge company or software project, and that's very understandable.

My main motivation for proposing this is to get everything done early on in the project so we don't have to worry about it in the future.

Thanks for hearing me out, by the way. I don't think this issue should be a huge priority at the moment, just something to keep in mind. I'll help whenever I can, but I can't promise too much since real life can be quite unstable 😅

**Comment by fstirlitz** [comment]
> As I understand, adding a COPYING.txt to the releases is sufficient. Could you help me write one?

This is not about just about applying the correct magic incantation to make lawyers go away. The big thing about using GPL code is that if you distribute binaries, you must also distribute source code. So you if you have CI generate your binaries (if I recall correctly, this is the case here), you should probably also have it generate a complete source package (no further downloads required), verify that you can actually build the same binary from that source code and offer that package for download as well. You may also go the ‘written offer’ route and simply say that anyone who asks for source code for this binary can simply ask you for it. But I imagine you’re not really willing to serve such requests.

In practice I don’t expect the authors of mutagen to go after you, but those are the licence requirements.

**Comment by pukkandan** [comment]
@fstirlitz 

> d) Convey the object code by offering access from a designated place (gratis or for a charge), and offer equivalent access to the Corresponding Source in the same way through the same place at no further charge. You need not require recipients to copy the Corresponding Source along with the object code. If the place to copy the object code is a network server, the Corresponding Source may be on a different server (operated by you or a third party) that supports equivalent copying facilities, provided you maintain clear directions next to the object code saying where to find the Corresponding Source. Regardless of what server hosts the Corresponding Source, you remain obligated to ensure that it is available for as long as needed to satisfy these requirements.

If I understand this correctly, providing a link pointing to the necessary dependancies is sufficient. Nowhere does it state that all the source code must be available as a single package.

**Comment by fstirlitz** [comment]
It doesn’t require packaging all dependencies with the source code, but it still has to meaningfully be *you* who is offering them for download. You can’t just say ‘oh well, you can get the dependencies from those other people’ and count on them not pulling a left-pad on you. The last sentence is clear here:

> Regardless of what server hosts the Corresponding Source, you remain obligated to ensure that it is available for as long as needed to satisfy these requirements.

Putting everything into a single, complete source code package just seems like the easiest way to satisfy this requirement.

Unless you really prefer the ‘written offer’ alternative. To be honest, it starts looking more and more attractive now…

**Comment by MrRawes** [comment]
actually under copyright if you break the license you cannot use it anymore unless the copyright holder asks (i mean EVERY copyright holder) or the license has some forgivenes clause for example in [GPLV3](https://spdx.org/licenses/GPL-3.0-only.html), [GPLV2](https://spdx.org/licenses/GPL-2.0-only.html) with [GPL-CC-1.0](https://spdx.org/licenses/GPL-CC-1.0.html) or [BlueOak-1.0.0](https://spdx.org/licenses/BlueOak-1.0.0.html)

**Comment by JarbasAl** [comment]
In our projects we have strict requirements and our CI pipeline tests for license validity, our tests fail when we add yt-dlp to our dependencies due to requirements.txt including mutagen

i went trough the code and confirmed mutagen is fully optional and can even be uninstalled, nevertheless i would like to ask that you consider removing it from requirements.txt or at least from the pypi package if possible. 

For my personal use case I whitelisted mutagen in our automations, but that also means we won't catch it if any of our packages decides to use mutagen, i understand this is low priority for the project but it is a pain point for us so i thought I should bring it up

Thanks!

**Comment by pukkandan** [comment]
All our dependencies are optional. You can install just yt-dlp with `pip install yt-dlp --no-deps`.

**Comment by JarbasAl** [comment]
> All our dependencies are optional. You can install just yt-dlp with `pip install yt-dlp --no-deps`.

but i can not have a project that depends on it and adds it to requirements.txt/setup.py , as that will drag mutagen, or can i specify --no-deps per package in a requirements.txt file?

as is i need to require users to install yt-dlp after the fact to dance around this

it's ok if this is not a concern and you dont care about it, dont mean to be pushy just to clarify the issue from my side

**Comment by bashonly** [comment]
@JarbasAl Since 87884f15580910e4e0fe0e1db73508debc657471 this should no longer be an issue
